File: packages/types/src/analytics.ts
import { z } from 'zod';

/** ABRO_PRD.md §26 "Monthly" -- GET /analytics/monthly?year=&month= */
export const monthlyAnalyticsQuerySchema = z.object({
  year: z.coerce.number().int().min(2000).max(2100),
  month: z.coerce.number().int().min(1).max(12),
});
export type MonthlyAnalyticsQuery = z.infer<typeof monthlyAnalyticsQuerySchema>;

/** ABRO_PRD.md §26 "Yearly" -- GET /analytics/yearly?year= */
export const yearlyAnalyticsQuerySchema = z.object({
  year: z.coerce.number().int().min(2000).max(2100),
});
export type YearlyAnalyticsQuery = z.infer<typeof yearlyAnalyticsQuerySchema>;

export interface CategoryAmount {
  category: string;
  amount: bigint;
}

/**
 * Shared field definitions (ABRO_PRD.md §26):
 *   Total Spending  -- sum of relevant expense totals (each expense counted
 *                      once, full amount, for every expense the user paid
 *                      or participated in).
 *   Your Contribution -- amount actually paid by the user.
 *   Your Share      -- amount allocated to the user.
 *   Net Position    -- Your Contribution minus Your Share (matches
 *                      BalancesService.getGroupSummary's netBalance).
 * All exclude soft-deleted expenses and SETTLEMENT-type rows (settlements
 * are reported separately, in `settlements`, to avoid double-counting a
 * debt transfer as new "spending").
 */
export interface MonthlyAnalytics {
  year: number;
  month: number;
  totalSpending: bigint;
  yourContribution: bigint;
  yourShare: bigint;
  netPosition: bigint;
  amountOwed: bigint;
  amountReceived: bigint;
  settlements: { paid: bigint; received: bigint };
  categoryBreakdown: CategoryAmount[];
}

export interface MonthlySpending {
  month: number;
  totalSpending: bigint;
}

export interface GroupSpending {
  groupId: string;
  groupName: string;
  totalSpending: bigint;
}

export interface YearlyAnalytics {
  year: number;
  yearlyTotal: bigint;
  monthlyTrend: MonthlySpending[];
  categoryDistribution: CategoryAmount[];
  groupSpending: GroupSpending[];
  personalContribution: bigint;
}

File: apps/api/internal/apitypes/analytics.go
package apitypes

// CategoryAmount / MonthlySpending / GroupSpending / MonthlyAnalytics /
// YearlyAnalytics mirror packages/types' analytics.ts response shapes.
// Amounts are wire strings, same convention as everywhere else.

type CategoryAmount struct {
	Category string `json:"category"`
	Amount   string `json:"amount"`
}

type MonthlySpending struct {
	Month         int    `json:"month"`
	TotalSpending string `json:"totalSpending"`
}

type GroupSpending struct {
	GroupID       string `json:"groupId"`
	GroupName     string `json:"groupName"`
	TotalSpending string `json:"totalSpending"`
}

type SettlementsSummary struct {
	Paid     string `json:"paid"`
	Received string `json:"received"`
}

type MonthlyAnalytics struct {
	Year              int                `json:"year"`
	Month             int                `json:"month"`
	TotalSpending     string             `json:"totalSpending"`
	YourContribution  string             `json:"yourContribution"`
	YourShare         string             `json:"yourShare"`
	NetPosition       string             `json:"netPosition"`
	AmountOwed        string             `json:"amountOwed"`
	AmountReceived    string             `json:"amountReceived"`
	Settlements       SettlementsSummary `json:"settlements"`
	CategoryBreakdown []CategoryAmount   `json:"categoryBreakdown"`
}

type YearlyAnalytics struct {
	Year                 int               `json:"year"`
	YearlyTotal          string            `json:"yearlyTotal"`
	MonthlyTrend         []MonthlySpending `json:"monthlyTrend"`
	CategoryDistribution []CategoryAmount  `json:"categoryDistribution"`
	GroupSpending        []GroupSpending   `json:"groupSpending"`
	PersonalContribution string            `json:"personalContribution"`
}
